### PR TITLE
Reworked `Ask<T>` to allocate fewer objects, `await` less

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Re-working the internals of `Ask<T>` to be more type safe, to `await` less internally, and to allocate fewer objects.

## Before
``` ini

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19041.1052 (2004/May2020Update/20H1)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=5.0.203
  [Host]     : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT
  DefaultJob : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT


```
|                        Method | Iterations |      Mean |     Error |    StdDev |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |----------- |----------:|----------:|----------:|----------:|------:|------:|----------:|
| RequestResponseActorSelection |      10000 | 83.313 ms | 0.7553 ms | 0.7065 ms | 4666.6667 |     - |     - |     19 MB |
|          CreateActorSelection |      10000 |  5.572 ms | 0.1066 ms | 0.1140 ms |  953.1250 |     - |     - |      4 MB |

## After (so far)
``` ini

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19041.1052 (2004/May2020Update/20H1)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=5.0.203
  [Host]     : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT
  DefaultJob : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT


```
|                        Method | Iterations |      Mean |     Error |    StdDev |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |----------- |----------:|----------:|----------:|----------:|------:|------:|----------:|
| RequestResponseActorSelection |      10000 | 71.216 ms | 0.9885 ms | 0.9246 ms | 4285.7143 |     - |     - |     17 MB |
|          CreateActorSelection |      10000 |  5.462 ms | 0.0495 ms | 0.0439 ms |  953.1250 |     - |     - |      4 MB |

### CODE REVIEW

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt [hunk 1]
```diff
@@ -864,9 +864,9 @@ namespace Akka.Actor
         public Failures() { }
         public System.Collections.Generic.List<Akka.Actor.Failure> Entries { get; }
     }
-    public class FutureActorRef : Akka.Actor.MinimalActorRef
+    public class FutureActorRef<T> : Akka.Actor.MinimalActorRef
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt:867]: Technically a breaking API change, but `FutureActorRef` isn't used as part of the public API. It's only used internally by `Ask<T>` (and maybe one other odd thing here or there)

File: src/core/Akka/Actor/ActorRef.cs [hunk 2]
```diff
@@ -131,7 +131,15 @@ protected override void TellInternal(object message, IActorRef sender)
...
                 {
-                    _result.TrySetResult(message);
+                    if (message is T t)
+                    {
+                        _result.TrySetResult(t);
+                    }
+                    else
+                    {
+                        _result.TrySetException(new ArgumentException(
+                            $"Received message of type [{message.GetType()}] - Ask expected message of type [{typeof(T)}]"));
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorRef.cs:141]: In the past this error was raised as an `InvalidCastException` - changed to `ArgumentException`

File: src/core/Akka/Actor/ActorRef.cs [hunk 3]
```diff
@@ -131,7 +131,15 @@ protected override void TellInternal(object message, IActorRef sender)
             {
                 if (Interlocked.Exchange(ref status, COMPLETED) == INITIATED)
                 {
-                    _result.TrySetResult(message);
+                    if (message is T t)
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorRef.cs:134]: In the past we casted from `object` to `T` - now we explicitly check for the presence of `T` and set the result using it.

This _should_ work with types that are assignable to `T` as well.

File: src/core/Akka/Actor/Futures.cs [hunk 4]
```diff
@@ -117,66 +117,44 @@ public static Task<T> Ask<T>(this ICanTell self, object message, TimeSpan? timeo
...
-        
-        private static async Task<object> Ask(ICanTell self, Func<IActorRef, object> messageFactory, IActorRefProvider provider,
-            TimeSpan? timeout, CancellationToken cancellationToken)
-        {
-            TaskCompletionSource<object> result = TaskEx.NonBlockingTaskCompletionSource<object>();
+            var result = TaskEx.NonBlockingTaskCompletionSource<T>();
 
             CancellationTokenSource timeoutCancellation = null;
             timeout = timeout ?? provider.Settings.AskTimeout;
-            var ctrList = new List<CancellationTokenRegistration>(2);
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Futures.cs:157]: Removed `List<T>` allocation and replaced with two `Nulalble<CancellationTokenRegistration>`

File: src/core/Akka/Actor/Futures.cs [hunk 5]
```diff
@@ -117,66 +117,44 @@ public static Task<T> Ask<T>(this ICanTell self, object message, TimeSpan? timeo
         /// This exception is thrown if the system can't resolve the target provider.
         /// </exception>
         /// <returns>TBD</returns>
-        public static async Task<T> Ask<T>(this ICanTell self, Func<IActorRef,object> messageFactory, TimeSpan? timeout, CancellationToken cancellationToken)
+        public static async Task<T> Ask<T>(this ICanTell self, Func<IActorRef, object> messageFactory, TimeSpan? timeout, CancellationToken cancellationToken)
         {
             await SynchronizationContextManager.RemoveContext;
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Futures.cs:122]: Not sure if this is needed - can tell the CTRs and everything else not to run in synchronization context if necessary.

File: src/core/Akka/Actor/Futures.cs [hunk 6]
```diff
@@ -192,17 +170,30 @@ internal static IActorRefProvider ResolveProvider(ICanTell self)
...
+                timeoutCancellation?.Dispose();
             }
         }
+
+        /// <summary>
+        /// Resolves <see cref="IActorRefProvider"/> for Ask pattern
+        /// </summary>
+        /// <param name="self">Reference to someone we are sending Ask request to</param>
+        /// <returns>Provider used for Ask pattern implementation</returns>
+        internal static IActorRefProvider ResolveProvider(ICanTell self)
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Futures.cs:184]: Method has not changed - just moved it.

**Review by Aaronontheweb — COMMENTED** [review]
Described some of my changes

**Comment by Aaronontheweb** [comment]
Shouldn't break wire compatibility, btw - `FutureActorRef` is converted to a `RemoteActorRef` when going over the wire. 

This will show up in the test suite if it's an issue though.

**Comment by Aaronontheweb** [comment]
Results after inlining the `SynchronizationContextManager`:

``` ini

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19041.1052 (2004/May2020Update/20H1)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=5.0.203
  [Host]     : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT
  DefaultJob : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT


```
|                        Method | Iterations |      Mean |     Error |    StdDev |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |----------- |----------:|----------:|----------:|----------:|------:|------:|----------:|
| RequestResponseActorSelection |      10000 | 71.921 ms | 0.7944 ms | 0.6634 ms | 4142.8571 |     - |     - |     17 MB |
|          CreateActorSelection |      10000 |  5.401 ms | 0.0458 ms | 0.0406 ms |  953.1250 |     - |     - |      4 MB |

Pretty much the same throughput, but Gen 1 GC dropped by ~5% or so.

File: src/core/Akka/Actor/Futures.cs [hunk 7]
```diff
@@ -117,92 +117,86 @@ public static Task<T> Ask<T>(this ICanTell self, object message, TimeSpan? timeo
...
             provider.RegisterTempActor(future, path);
             var message = messageFactory(future);
             self.Tell(message, future);
+            var prevContext = SynchronizationContext.Current;
 
+  
             try
             {
+                SynchronizationContext.SetSynchronizationContext(null);
                 return await result.Task;
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Futures.cs:168]: Can't get rid of this `await` unfortunately - it's much slower to do all of the cleanup using a `ContinueWith` than it is to just `await` the task and do cleanup in a `finally` block instead.

**Comment by Aaronontheweb** [comment]
Next results - removed final `await` call and an unnecessary delegate allocation:

``` ini

BenchmarkDotNet=v0.13.0, OS=Windows 10.0.19041.1052 (2004/May2020Update/20H1)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=5.0.203
  [Host]     : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT
  DefaultJob : .NET Core 3.1.15 (CoreCLR 4.700.21.21202, CoreFX 4.700.21.21402), X64 RyuJIT


```
|                        Method | Iterations |      Mean |     Error |    StdDev |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------ |----------- |----------:|----------:|----------:|----------:|------:|------:|----------:|
| RequestResponseActorSelection |      10000 | 63.210 ms | 0.9977 ms | 0.9332 ms | 3875.0000 |     - |     - |     15 MB |
|          CreateActorSelection |      10000 |  5.502 ms | 0.0470 ms | 0.0417 ms |  953.1250 |     - |     - |      4 MB |

**Review by Arkatufus — APPROVED** [review]
LGTM

**Pull request merged**